http4s-circe/src/test/scala/io/isomarcte/errors4s/http4s/circe/middleware/client/PassthroughCirceHttpErrorTest.scala

```diff
@@ -0,0 +1,143 @@
+
+  "A response with a application/problem+json Content-Type" should "be raise as an error in F if it is well formed" in
+    sio {
+      val client: Client[SyncIO] = constClientF(httpErrorAsResponse[SyncIO](testError))
+
+      client
+        .run(testRequest)
+        .use[SyncIO, Unit](resp => SyncIO(fail(s"Expected error, but got $resp")))
+        .attemptT
+        .foldF(
+          e =>
+            e match {
+              case e: ExtensibleCirceHttpProblem =>
+                SyncIO(e.asJson shouldBe testError.asJson)
+              case otherwise =>
+                SyncIO(fail(s"Expected ExtensibleCirceHttpProblem, but got $otherwise"))
+            },
+          value => SyncIO(fail(s"Expected error, but got successful $value"))
+        )
+    }
+
+  it should "yield the original response if parsing fails (e.g. the Content-Type is wrong)" in
+    sio {
+      val client: Client[SyncIO] = constClient(responseWithContentType)
+
+      client.run(testRequest).use[SyncIO, Unit](resp => compareResponses(resp, responseWithContentType))
+    }
+
+  "The PassthroughCirceHttpError middleware" should
+    "only read the body if there is an application/problem+json, and in that case it should only do so once" in
+    sio {
+      Ref
+        .of[SyncIO, Int](0)
+        .flatMap { ref =>
+          val responseBody: String = "Body"
+          val response: Response[SyncIO] = Response(
+            status = Status.Ok,
+            headers = Headers.of(`Content-Type`(MediaType.application.`problem+json`)),
+            body = Stream
+              .evalUnChunk(ref.update(_ + 1) *> SyncIO.pure(Chunk.bytes(responseBody.getBytes(StandardCharsets.UTF_8))))
+          )
+
+          val client: Client[SyncIO] = constClient(response)
+
+          // The result should be _2_ and not _1_ because the body is evaluated
+          // once in the middleware and once again by the `compareResponses`
+          // helper (which uses the original `fs2.Stream` value.
+          client.run(testRequest).use(value => compareResponses(value, response)) *>
+            ref.get.flatMap(value => SyncIO(value shouldBe 2))
+        }
+    }
+
+  it should "not read the body at all if the content-type is not an application/problem+json" in
+    sio {
+      Ref
+        .of[SyncIO, Int](0)
+        .flatMap { ref =>
+          val responseBody: String = "Body"
+          val response: Response[SyncIO] = Response(
+            status = Status.Ok,
+            body = Stream
+              .evalUnChunk(ref.update(_ + 1) *> SyncIO.pure(Chunk.bytes(responseBody.getBytes(StandardCharsets.UTF_8))))
+          )
+
+          val client: Client[SyncIO] = constClient(response)
+
+          client
+            .run(testRequest)
+            .use(value =>
+              // Throw away the body, which should prevent the Ref from ever being
+              // updated if the middleware is working correctly.
+              SyncIO.pure(value.copy(body = Stream.empty))
```

Comment:
Should this check instead to see if the body is returned as expected?

Reply:
Not in this test. The only want to be sure the middleware didn't touch the body at all is to also have the happy path code not touch the body at all. If the test pulls out the body then it will trigger the Ref, and we can't be sure if it was the middleware or the test code which triggered the ref to update.